c_src/olm_nif.c

```diff
@@ -0,0 +1,328 @@
+#include "erl_nif.h"
+#include <olm/inbound_group_session.h>
+#include <olm/olm.h>
+#include <olm/outbound_group_session.h>
+#include <olm/pk.h>
+#include <olm/sas.h>
+
+static ErlNifResourceType* account_resource;
+
+void
+account_dtor(ErlNifEnv* caller_env, void* account)
+{
+    olm_clear_account(account);
+}
+
+static int
+nif_load(ErlNifEnv* env, void** priv_data, ERL_NIF_TERM load_info)
+{
+    int flags        = ERL_NIF_RT_CREATE | ERL_NIF_RT_TAKEOVER;
+    account_resource = enif_open_resource_type(
+        env, NULL, "account", account_dtor, flags, NULL);
+
+    return 0;
+}
+
+static ERL_NIF_TERM
+version(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    uint8_t major, minor, patch;
+    olm_get_library_version(&major, &minor, &patch);
+
+    return enif_make_tuple3(env,
+                            enif_make_uint(env, major),
+                            enif_make_uint(env, minor),
+                            enif_make_uint(env, patch));
+}
+
+static ERL_NIF_TERM
+account_last_error(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    OlmAccount* account;
+    enif_get_resource(env, argv[0], account_resource, (void**) &account);
+
+    // Perhaps make this atoms?
+    const char* last_error = olm_account_last_error(account);
+
+    return enif_make_string(env, last_error, ERL_NIF_LATIN1);
+}
+
+static ERL_NIF_TERM
+create_account(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    size_t account_size = olm_account_size();
+
+    // Allocate memory based on account_size.
+    OlmAccount* memory  = enif_alloc_resource(account_resource, account_size);
+    OlmAccount* account = olm_account(memory);
+
+    size_t random_length = olm_create_account_random_length(account);
+    char   bytes[random_length];
+
+    size_t result = olm_create_account(account, bytes, random_length);
+
+    // Return {:ok, account_ref} or {:error, last_error}.
+    if (result == olm_error()) {
+        ERL_NIF_TERM error_atom    = enif_make_atom(env, "error");
+        ERL_NIF_TERM error_message = enif_make_string(
+            env, olm_account_last_error(account), ERL_NIF_LATIN1);
+
+        return enif_make_tuple2(env, error_atom, error_message);
+    } else {
```

Comment:
I would remove the else. The if will short circuit the expression with the return statement anyways.